extends Node2D

var BattleScene = preload("./Battle.tscn")
var GameOverScene = preload("./game_over/GameOver.tscn")
var DialogueScene = preload("res://components/dialogue/Dialogue.tscn")

func _ready():
	$Mask.visible = false
	$CanvasLayer/Bg.visible = true
	
	$CanvasLayer/Bg.visible = false
	var test_battle = self.create_battle(Constants.CREATURES.troll, 
	[
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.crash,
		Constants.ELEMENTS.crash
	])
	yield(test_battle, "battle_finished")
	
	yield(self.show_dialogue([{
		"type": "sentence",
		"content": "ä½ å«é‡‘,æ˜¯ä¸€ä¸ªå…ƒç´ ç‚¼æˆå¸ˆ."
	}, {
		"type": "sentence",
		"content": "æ‰€è°“å…ƒç´ ç‚¼æˆå¸ˆ,å°±æ˜¯èƒ½å¤ŸæŠŠä¸–ç•Œä¸­çš„ç‰©è´¨å’Œç²¾ç¥ç»„åˆèµ·æ¥çš„äºº."
	}, {
		"type": "sentence",
		"content": "æœ€è¿‘ä¼ é—»ä¸–ç•Œä¸Šä¸€ä¸ªæ–°çš„é­”ç‹å‡ºç°,ä¼—å¤šé­”ç‰©æ¶Œç°."
	}, {
		"type": "sentence",
		"content": "ä¸ºäº†ä¸–ç•Œå’Œå¹³,ä½ å†³å®šå‰å¾€é­”ç‹çš„åŸå ¡å»æ‰“å€’é­”ç‹."
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "èƒ½å¤Ÿæ‰“å€’é­”ç‹çš„,åªæœ‰ kono é‡‘å“’!"
	}]), "completed")
	
	yield(self.show_dialogue([{
		"type": "sentence",
		"content": "æ¥åˆ°é­”ç‹çš„åŸå ¡å,ä¸€åªå²è±å§†æŒ¡åœ¨ä½ çš„é¢å‰!"
	}]), "completed")
	$CanvasLayer/Bg.visible = false
	var slime_battle = self.create_battle(Constants.CREATURES.slime,
	[
		Constants.ELEMENTS.fire, 
		Constants.ELEMENTS.concentrated
	])
	yield(self.show_dialogue([{
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "å²è±å§†å—,æˆ‘å°è±¡ä¸­è¿™ç§ç”Ÿç‰©æ¯”è¾ƒæ€•ç«."
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "åˆæˆä¸€äº›ç«ç›¸å…³çš„æŠ€èƒ½åº”è¯¥å°±èƒ½æ‰“å€’å®ƒ."
	}]), "completed")
	var slime_win = yield(slime_battle, "battle_finished")
	if !slime_win:
		self.show_game_over()
		return
	
	$CanvasLayer/Bg.visible = true
	yield(self.show_dialogue([{
		"type": "sentence",
		"content": "æ‰“è´¥å²è±å§†å,ä½ ç»§ç»­å‰è¿›."
	}, {
		"type": "sentence",
		"content": "åˆšèµ°äº†æ²¡å¤šä¹…,ä¸€åªéª·é«…ä»æš—å¤„å‘ä½ è¢­å‡»!"
	}]), "completed")
	$CanvasLayer/Bg.visible = false
	var skeleton_battle = self.create_battle(Constants.CREATURES.skleton, 
	[
		Constants.ELEMENTS.fire, 
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.crash
	])
	yield(self.show_dialogue([{
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "å””,éª·é«…è¿™ç§ç”Ÿç‰©ï¼Œç”¨ç«çƒ§è¿˜æ˜¯ç”¨æ°´æ³¡å¯¹å®ƒéƒ½æ²¡ä»€ä¹ˆå½±å“."
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "æœ€å¥½çš„æ–¹å¼æ˜¯è®©å®ƒç²‰èº«ç¢éª¨,æŠŠå®ƒéª¨ç°éƒ½æ‰¬äº†!æˆ–è€…ä¹Ÿå¯ä»¥ç”¨æ›´åŠ çŒ›çƒˆçš„ç«ç„°çƒ§åˆ°åªå‰©ä¸‹éª¨ç°."
	}]), "completed")
	var skeleton_win = yield(skeleton_battle, "battle_finished")
	if !skeleton_win:
		self.show_game_over()
		return
	
	$CanvasLayer/Bg.visible = true
	yield(self.show_dialogue([{
		"type": "sentence",
		"content": "æ‰“è´¥æ€ªç‰©åä½ ç»§ç»­å‰è¿›."
	}, {
		"type": "sentence",
		"content": "çªç„¶ä¸€ä¸ªå·¨å¤§çš„èº«å½±å‡ºç°åœ¨ä½ é¢å‰---æ˜¯å·¨é­”.å®ƒå‘ä½ å¼ ç‰™èˆçˆªæ‰‘è¿‡æ¥!"
	}]), "completed")
	$CanvasLayer/Bg.visible = false
	var troll_battle = self.create_battle(Constants.CREATURES.troll, 
	[
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.crash,
		Constants.ELEMENTS.crash
	])
	yield(self.show_dialogue([{
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "æˆ‘è§‰å¾—ä½ åº”è¯¥éƒ½ç†Ÿæ‚‰è¯¥æ€ä¹ˆæˆ˜æ–—äº†å§,é‚£æ¥ä¸‹æ¥æˆ‘å°±ä¸ç”¨å†å‡è£…è‡ªç„¶è‡ªè¯­ç»™ä½ æç¤ºäº†!"
	}]), "completed")
	var troll_win = yield(troll_battle, "battle_finished")
	if !troll_win:
		self.show_game_over()
		return
	$CanvasLayer/Bg.visible = true
	
	yield(self.show_dialogue([{
		"type": "sentence",
		"content": "ä½ å·²ç»å¾ˆæ¥è¿‘é­”ç‹çš„ä½ç½®äº†.å½“ä½ èµ°åˆ°é­”ç‹çš„é—¨å‰,ç‚é­”å‡ºç°åœ¨ä½ é¢å‰."
	}, {
		"type": "sentence",
		"actor": "ç‚é­”",
		"content": "ç«™ä½!æƒ³è¦æ‰“å€’é­”ç‹çš„è¯å°±è¦å…ˆæ‰“è´¥æˆ‘!"
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "æˆ‘è¿˜ä»¥ä¸ºé™¤äº†é­”ç‹çš„æ€ªç‰©éƒ½ä¸ä¼šè¯´è¯å‘¢!æ—¢ç„¶ä½ æ±‚æˆ‘æ‰“è´¥ä½ ,é‚£æˆ‘å°±å‹‰ä¸ºå…¶éš¾å’Œä½ æ‰“ä¸€åœºå§."
	}]), "completed")
	$CanvasLayer/Bg.visible = false
	var balrog_battle = self.create_battle(Constants.CREATURES.balrog, 
	[
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.crash,
		Constants.ELEMENTS.crash
	])
	var balrog_win = yield(balrog_battle, "battle_finished")
	if !balrog_win:
		self.show_game_over()
		return
	$CanvasLayer/Bg.visible = true
	yield(self.show_dialogue([{
		"type": "sentence",
		"actor": "ç‚é­”",
		"content": "...å‘µå‘µ,å³ä½¿ä½ èƒ½å¤Ÿæ‰“è´¥æˆ‘,ä½ ä¹Ÿæ²¡å¯èƒ½æ‰“è´¥é­”ç‹."
	}, {
		"type": "sentence",
		"actor": "ç‚é­”",
		"content": "é­”ç‹æ˜¯æ²¡æœ‰ä»»ä½•å¼±ç‚¹çš„!"
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "æ²¡æœ‰å¼±ç‚¹å—,è¿™è¿˜æ˜¯æŒºæ£˜æ‰‹çš„,åªèƒ½æ‰“æ¶ˆè€—æˆ˜å—."
	}]), "completed")
	
	yield(self.show_dialogue([{
		"type": "sentence",
		"content": "ä½ ç»ˆäºæ¥åˆ°é­”ç‹é¢å‰,é­”ç‹çœ‹åˆ°äº†ä½ é‚ªé­…ä¸€ç¬‘."
	}, {
		"type": "sentence",
		"actor": "é­”ç‹",
		"content": "ä½ ç»ˆäºæ¥åˆ°è¿™é‡Œäº†,æˆ‘è¿˜ä»¥ä¸ºè¿™ä¸ªä¸–ç•Œçš„äººéƒ½æ˜¯èƒ†å°é¬¼,æ²¡æƒ³åˆ°è¿˜æœ‰äººæ•¢è¿‡æ¥."
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "åºŸè¯å°‘è¯´,çœ‹æˆ‘è¿‡æ¥æŠŠä½ ç—›æä¸€é¡¿!"
	}, {
		"type": "sentence",
		"actor": "é­”ç‹",
		"content": "éä½†ä¸é€ƒèµ°,è¿˜å‘æˆ‘è¿™è¾¹èµ°æ¥å—.æ˜æ˜æˆ‘çš„å®ˆå«éƒ½å‘Šè¯‰è¿‡ä½ æˆ‘æœ‰å¤šå¯æ€•äº†."
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "ä¸èµ°í ¼í¸²è¿‘ç‚¹çš„è¯,æ€ä¹ˆç—›æä½ ä¸€é¡¿å•Š!"
	}, {
		"type": "sentence",
		"actor": "é­”ç‹",
		"content": "éœéœ,é‚£ä½ å°±å†èµ°è¿‘ç‚¹å¥½äº†."
	}, {
		"type": "sentence",
		"content": "ä½ å’Œé­”ç‹äº’ç›¸èµ°å‘å¯¹æ–¹,ä½ ä»¬ä¹‹å‰å‘å‡ºä¸¤è‚¡å¼ºå¤§çš„æ°”æµç›¸äº’ç¢°æ’,æˆ˜æ–—å¼€å§‹!"
	}]), "completed")
	$CanvasLayer/Bg.visible = false
	var king_battle = self.create_battle(Constants.CREATURES.king, 
	[
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.fire,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.water,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.mud,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.concentrated,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.warm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.firm,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.wrath,
		Constants.ELEMENTS.crash,
		Constants.ELEMENTS.crash
	])
	var king_win = yield(king_battle, "battle_finished")
	if !king_win:
		self.show_game_over()
		return
	
	$CanvasLayer/Bg.visible = true
	yield(self.show_dialogue([{
		"type": "sentence",
		"actor": "é­”ç‹",
		"content": "ä¸å¯èƒ½!æˆ‘ç«Ÿç„¶ä¼šè¾“!"
	}, {
		"type": "sentence",
		"actor": "é­”ç‹",
		"content": "ä¸è¿‡æˆ‘è¿˜æ²¡å¤±è´¥,åªè¦èƒ½å»é‚£ä¸ªåœ°æ–¹,æ¶ˆç­ä½ æ˜¯è½»è€Œæ˜“ä¸¾çš„äº‹..."
	}, {
		"type": "sentence",
		"actor": "é‡‘",
		"portrait": "res://assets/jin.png",
		"content": "48å°æ—¶çˆ†è‚æ¸¸æˆåªèƒ½åšåˆ°è¿™é‡Œäº†,ä½ è¿˜æƒ³äºŒé˜¶æ®µå˜èº«?"
	}, {
		"type": "sentence",
		"content": "ä½ ç»™äº†é­”ç‹æœ€åä¸€å‡»,é­”ç‹è¢«æ‰“å€’äº†."
	}, {
		"type": "sentence",
		"content": "ä»æ­¤ä¸–ç•Œæ¢å¤å’Œå¹³,å¯å–œå¯è´º."
	}]), "completed")
	
func on_battle_finished(win, node: Node):
	node.queue_free()
	yield(node, "tree_exited")
		
func show_game_over():
	var game_over = self.GameOverScene.instance()
	self.add_child(game_over)
	
func create_battle(enemy, elements):
	var battle = self.BattleScene.instance()
	battle.enemy = enemy
	battle.state.elements = elements
	battle.connect("battle_finished", self, "on_battle_finished", [battle])
	self.add_child(battle)
	return battle

func show_dialogue(dialogue):
	$Mask.visible = true
	var node = self.DialogueScene.instance()
	$Popups.add_child(node)
	yield(node.show_dialogue(dialogue), "completed")
	$Mask.visible = false
